hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMBatchKeyDeleteRequest.java

```diff
@@ -0,0 +1,229 @@
+    OzoneManagerProtocolProtos.UserInfo userInfo =
+        getOmRequest().getUserInfo();
+
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+
+    try {
+      for (KeyArgs deleteKeyArgs : deleteKeyArgsList) {
+        volumeName = deleteKeyArgs.getVolumeName();
+        bucketName = deleteKeyArgs.getBucketName();
+        keyName = deleteKeyArgs.getKeyName();
+        auditMap = buildKeyArgsAuditMap(deleteKeyArgs);
+
+        // check Acl
+        checkKeyAcls(ozoneManager, volumeName, bucketName, keyName,
+            IAccessAuthorizer.ACLType.DELETE, OzoneObj.ResourceType.KEY);
+
+        String objectKey = omMetadataManager.getOzoneKey(
+            volumeName, bucketName, keyName);
+
+        acquiredLock = omMetadataManager.getLock().acquireWriteLock(BUCKET_LOCK,
+            volumeName, bucketName);
+        // Validate bucket and volume exists or not.
+        validateBucketAndVolume(omMetadataManager, volumeName, bucketName);
+
+        OmKeyInfo omKeyInfo = omMetadataManager.getKeyTable().get(objectKey);
+        if (omKeyInfo == null) {
+          throw new OMException("Key not found", KEY_NOT_FOUND);
+        }
+
+        // Check if this transaction is a replay of ratis logs.
+        if (isReplay(ozoneManager, omKeyInfo, trxnLogIndex)) {
+          // Replay implies the response has already been returned to
+          // the client. So take no further action and return a dummy
+          // OMClientResponse.
+          throw new OMReplayException();
+        }
+
+        // Set the UpdateID to current transactionLogIndex
+        omKeyInfo.setUpdateID(trxnLogIndex, ozoneManager.isRatisEnabled());
+
+        // Update table cache.
+        omMetadataManager.getKeyTable().addCacheEntry(
+            new CacheKey<>(omMetadataManager.getOzoneKey(
+                volumeName, bucketName, keyName)),
+            new CacheValue<>(Optional.absent(), trxnLogIndex));
+
+        // No need to add cache entries to delete table. As delete table will
+        // be used by DeleteKeyService only, not used for any client response
+        // validation, so we don't need to add to cache.
+        // TODO: Revisit if we need it later.
+
+        if (acquiredLock) {
+          omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+              bucketName);
+          acquiredLock = false;
+        }
+        omKeyInfoList.add(omKeyInfo);
+      }
+      omClientResponse = new OMBatchKeyDeleteResponse(omResponse
+          .setDeleteBatchKeyResponse(DeleteBatchKeyResponse.newBuilder())
+          .build(),
+          omKeyInfoList, ozoneManager.isRatisEnabled());
+      result = Result.SUCCESS;
+    } catch (IOException ex) {
+      if (ex instanceof OMReplayException) {
+        result = Result.REPLAY;
+        omClientResponse = new OMKeyDeleteResponse(createReplayOMResponse(
```

Comment:
Can we return the list of deleted keys together with the exception to client so that client knows which keyes are deleted and which key trigger the exception?